src/inmanta/server/agentmanager.py

```diff
@@ -806,6 +842,10 @@ async def _ensure_agents(self, env: data.Environment, agents: List[str], restart
             :param agents: A list of agent names that possibly should be started in this environment.
             :param restart: Restart all agents even if the list of agents is up to date.
         """
+        # silently ignore requests if this environment is halted
+        if env.halted:
+            return False
```

Comment:
This check should be executed in the `agent_lock` to prevent race conditions. Currently the following scenario is possible:

* This `env.halted` condition returns False and after that a context switch happens
* Environment gets halted by pressing the red button.
* Resume context bullet 1: __do_start_agent() is called.